nmagent/internal/retry.go

```diff
@@ -0,0 +1,119 @@
+package internal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"time"
+)
+
+const (
+	noDelay = 0 * time.Nanosecond
+)
+
+// TemporaryError is an error that can indicate whether it may be resolved with
+// another attempt
+type TemporaryError interface {
+	error
+	Temporary() bool
+}
+
+// Retrier is a construct for attempting some operation multiple times with a
+// configurable backoff strategy
+type Retrier struct {
+	Cooldown CooldownFactory
+}
+
+// Do repeatedly invokes the provided run function while the context remains
+// active. It waits in between invocations of the provided functions by
+// delegating to the provided Cooldown function
+func (r Retrier) Do(ctx context.Context, run func() error) error {
+	cooldown := r.Cooldown()
+
+	for {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		err := run()
+		if err != nil {
+			// check to see if it's temporary
+			var tempErr TemporaryError
+			if ok := errors.As(err, &tempErr); ok && tempErr.Temporary() {
+				delay, err := cooldown()
+				if err != nil {
+					return fmt.Errorf("sleeping during retry: %w", err)
+				}
+				time.Sleep(delay)
+				continue
+			}
+
+			// since it's not temporary, it can't be retried, so...
+			return err
+		}
+		return nil
```

Comment:
Good point, in practice it probably would be pretty close to immediate since the context is propagated down to the `http.(*Client)`, but that's not a good assumption for this to have (since it's supposed to work with anything). We're definitely converging on your implementation of retries, but with some functional programming and temporary errors :) We may want to have that discussion about retry logic in general... possibly move this into a top-level package, etc.

I'm also probably going to tweak this so that the function being retried passes down the context as well... it could break up the context heirarchy a bit if it gets passed in via the closure when you actually go to use this.